Sources/MlemMiddleware/API Client/ApiClient+Post.swift

```diff
@@ -68,15 +68,94 @@ extension ApiClient: PostFeedProvider {
         return nil
     }
     
+    /// Mark the given post as read. Works on all versions.
+    /// On v0.19.0 and above, calling this will also mark any queued posts as read unless `includeQueuedPosts` is set to `false`.
+    public func markPostAsRead(
+        id: Int,
+        read: Bool = true,
+        includeQueuedPosts: Bool = true,
+        semaphore: UInt? = nil
+    ) async throws {
+        // We *must* use `postId` in 0.18 versions, and we *must* use `postIds` from 0.19.4 onwards.
+        let version = try await version
+        let request: MarkPostAsReadRequest
+        if version >= .v19_0 {
+            try await self.markPostsAsRead(
+                ids: [id],
+                read: read,
+                includeQueuedPosts: includeQueuedPosts,
+                semaphore: semaphore
+            )
+        } else {
+            request = MarkPostAsReadRequest(postId: id, read: read, postIds: nil)
+            let response = try await perform(request)
+            if !response.success {
+                throw ApiClientError.unsuccessful
+            }
+            markReadQueue.remove(id)
+            if let post = caches.post2.retrieveModel(cacheId: id) {
+                post.readManager.updateWithReceivedValue(read, semaphore: semaphore)
+                post.readQueued = false
+            }
+        }
+    }
+    
+    /// Mark the given posts as read. Only works on v0.19.0 and above; on lower versions, use `markPostAsRead` instead.
+    /// Calling this will also mark any queued posts as read unless `includeQueuedPosts` is set to `false`.
+    public func markPostsAsRead(
+        ids: Set<Int>,
+        read: Bool = true,
+        includeQueuedPosts: Bool = true,
+        semaphore: UInt? = nil
+    ) async throws {
+        let version = try await version
+        guard version >= .v19_0 else { throw ApiClientError.unsupportedLemmyVersion }
+        
+        let idsToSend: Set<Int>
+        let markReadQueueCopy: Set<Int>
+        if read, includeQueuedPosts {
+            idsToSend = ids.union(self.markReadQueue)
+            markReadQueueCopy = self.markReadQueue
+            self.markReadQueue.removeAll()
```

Comment:
Not thread safe; if two threads execute L118 before either executes L119, the same request will be sent twice--hardly a cardinal sin, but could potentially result in server spam if
- `markReadQueue` has gotten long, leading to L118 executing slowly
- User is scrolling fast with mark read on scroll activated

This function should either be documented as thread-unsafe or perform thread safety checks internally.

Reply:
Would this be enough to fix it? Or do you think I should semaphore the entire method?

```swift
markReadQueueCopy = self.markReadQueue
self.markReadQueue.removeAll()
idsToSend = ids.union(markReadQueueCopy)
```

Reply:
Given just how easy it is to spin up a ton of threads all touching the underlying set (i.e., scroll fast with mark read on scroll), it's almost certain that we're going to create at best a data race and at worst a segfault unless we make sure that all of the data access to the underlying set is thread-safe.

What we should use--and, I realize now, what I should have been using all along instead of mucking around with semaphores like a C nerd--is an `actor`! Good ol' hackingwithswift has a [nice article](https://www.hackingwithswift.com/quick-start/concurrency/what-is-an-actor-and-why-does-swift-have-them) on them; as far as I can tell, the tl;dr is that everything inside an actor is executed on a serial queue, so we can just do something like:

```
actor MarkReadQueue {
  Set<Int> queuedPosts

  // clears queuedPosts and returns a set containing all removed posts
  func popAll() -> Set<Int> { ... }
}
```

Reply:
I didn't know about those! Thanks :D